drivers/client_access_control/app/controllers/client_access_control/history_controller.rb

```diff
@@ -57,7 +57,7 @@ def pdf
       @project_type_counts = @dates.values.flatten.group_by { |en| HUD.project_type en[:project_type] }.transform_values(&:count)
       file_name = 'service_history.pdf'
 
-      template_file = 'client_access_control/history/pdf'
+      template_file = File.join(Rails.root, "drivers/client_access_control/app/views/client_access_control/history/pdf.haml")
       pdf = nil
       PdfGenerator.new.perform(
         html: render(file: template_file, layout: false),
```

Comment:
Is this somehow different from other places we call `render(file: ...`.  It appears there are another 17 or so we may need to update if this is an important change.

Reply:
It does appear a change needs to be made.  Not yet sure what exactly it is.

Reply:
Dropping this here for future discovery.  It doesn't appear the api has changed for url_for
https://github.com/rails/rails/blob/157920aead96865e3135f496c09ace607d5620dc/actionpack/lib/action_dispatch/routing/polymorphic_routes.rb#L57
https://github.com/rails/rails/blob/5-2-stable/actionpack/lib/action_dispatch/routing/polymorphic_routes.rb
and delayed job doesn't appear to be the culprit, though it may be.
https://github.com/collectiveidea/delayed_job/blob/master/lib/delayed/lifecycle.rb

```
wrong number of arguments (given 2, expected 3)
/bundle/ruby/2.7.0/gems/actionview-6.1.4.4/lib/action_view/base.rb:230:in `initialize'
/bundle/ruby/2.7.0/gems/actionpack-6.1.4.4/lib/action_dispatch/routing/url_for.rb:108:in `initialize'
/app/app/models/grda_warehouse/document_exports/client_performance_export.rb:31:in `new'
/app/app/models/grda_warehouse/document_exports/client_performance_export.rb:31:in `view'
/app/app/models/grda_warehouse/document_exports/client_performance_export.rb:13:in `block in perform'
/app/app/models/concerns/document_export_behavior.rb:70:in `with_status_progression'
/app/app/models/grda_warehouse/document_exports/client_performance_export.rb:10:in `perform'
/app/app/models/concerns/document_export_job_behavior.rb:13:in `perform'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/execution.rb:48:in `block in perform_now'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:117:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/marginalia-1.11.1/lib/marginalia/railtie.rb:32:in `block (2 levels) in insert_into_active_job'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `instance_exec'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/i18n-1.8.11/lib/i18n.rb:314:in `with_locale'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/translation.rb:9:in `block (2 levels) in <module:Translation>'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `instance_exec'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/core_ext/time/zones.rb:66:in `use_zone'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/timezones.rb:9:in `block (2 levels) in <module:Timezones>'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `instance_exec'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/instrumentation.rb:21:in `block in instrument'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/notifications.rb:203:in `block in instrument'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/notifications.rb:203:in `instrument'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/instrumentation.rb:31:in `instrument'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/instrumentation.rb:14:in `block (2 levels) in <module:Instrumentation>'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `instance_exec'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/logging.rb:22:in `block in tag_logger'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/tagged_logging.rb:99:in `block in tagged'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/tagged_logging.rb:37:in `tagged'
/bundle/ruby/2.7.0/gems/logstop-0.2.7/lib/logstop/formatter.rb:17:in `method_missing'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/tagged_logging.rb:99:in `tagged'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/logging.rb:22:in `tag_logger'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/logging.rb:15:in `block (2 levels) in <module:Logging>'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `instance_exec'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:137:in `run_callbacks'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/execution.rb:47:in `perform_now'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/execution.rb:25:in `block in execute'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:117:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/railtie.rb:47:in `block (4 levels) in <class:Railtie>'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/execution_wrapper.rb:88:in `wrap'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/reloader.rb:72:in `block in wrap'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/execution_wrapper.rb:88:in `wrap'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/reloader.rb:71:in `wrap'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/railtie.rb:46:in `block (3 levels) in <class:Railtie>'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `instance_exec'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:126:in `block in run_callbacks'
/bundle/ruby/2.7.0/gems/activesupport-6.1.4.4/lib/active_support/callbacks.rb:137:in `run_callbacks'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/execution.rb:23:in `execute'
/bundle/ruby/2.7.0/gems/activejob-6.1.4.4/lib/active_job/queue_adapters/delayed_job_adapter.rb:42:in `perform'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/backend/base.rb:81:in `block in invoke_job'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:61:in `block in initialize'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:66:in `execute'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:40:in `run_callbacks'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/backend/base.rb:78:in `invoke_job'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:230:in `block (2 levels) in run'
/usr/local/lib/ruby/2.7.0/timeout.rb:95:in `block in timeout'
/usr/local/lib/ruby/2.7.0/timeout.rb:105:in `timeout'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:230:in `block in run'
/usr/local/lib/ruby/2.7.0/benchmark.rb:308:in `realtime'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:229:in `run'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:312:in `block in reserve_and_run_one_job'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:61:in `block in initialize'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:66:in `execute'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:40:in `run_callbacks'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:312:in `reserve_and_run_one_job'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:213:in `block in work_off'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:212:in `times'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:212:in `work_off'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:175:in `block (4 levels) in start'
/usr/local/lib/ruby/2.7.0/benchmark.rb:308:in `realtime'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:174:in `block (3 levels) in start'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:61:in `block in initialize'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:66:in `execute'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:40:in `run_callbacks'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:173:in `block (2 levels) in start'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:172:in `loop'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:172:in `block in start'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/plugins/clear_locks.rb:7:in `block (2 levels) in <class:ClearLocks>'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:79:in `block (2 levels) in add'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:61:in `block in initialize'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:79:in `block in add'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:66:in `execute'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/lifecycle.rb:40:in `run_callbacks'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/worker.rb:171:in `start'
/bundle/ruby/2.7.0/gems/delayed_job-4.1.9/lib/delayed/tasks.rb:9:in `block (2 levels) in <main>'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:281:in `block in execute'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:281:in `each'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:281:in `execute'
/app/config/initializers/rake.rb:6:in `execute'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:219:in `block in invoke_with_call_chain'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:199:in `synchronize'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:199:in `invoke_with_call_chain'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/task.rb:188:in `invoke'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:160:in `invoke_task'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:116:in `each'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:116:in `block in top_level'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:125:in `run_with_threads'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:110:in `top_level'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:83:in `block in run'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:186:in `standard_exception_handling'
/bundle/ruby/2.7.0/gems/rake-13.0.6/lib/rake/application.rb:80:in `run'
bin/rake:4:in `<main>'
```

Reply:
OK, tracked this down. 
https://blog.saeloun.com/2019/10/07/ruby-2-7-keyword-arguments-redesign.html
`def perform(export_id: nil)`
needs to be:
```
def perform(opts)
  export_id = opts[:export_id]
```
or the calling method needs to pass **{export_id: 1}.

There's probably a more systematic way to handle this, as is, this could be tricky to track down.  Basically, you can no longer pass a hash to a method that expects key word arguments.

Reply:
Hmm, this appears to be a bit of a red herring.  I believe the actual issue is the API for `ActionView::Base#initialize` is different or not quite getting called right.
`PdfExportTemplateBase.new(context, view_assigns, controller_class.new)` in `app/models/concerns/hud_reports/hud_pdf_export_concern.rb` is as close as I've been able to get.